### PR TITLE
LLM-20: support running hallucinations bechmarks and UNQOVER (bias dataset)

### PR DESCRIPTION
**Pull request by tomerr-hirundo**
medhallu, halueval and unqover supported

added a brief summary per model (bias and hallucinations with the score)

changed docs and main evaluate function to be more concise (user only has to input the model name and behavior type)

### CODE REVIEW

**Review by benglewis — COMMENTED** [review]
Mostly LGTM, just a few important comments / suggestions

File: README.md [hunk 1]
```diff
@@ -2,67 +2,122 @@
...
+# 3) Install dependencies using uv
+# (Assumes the `uv` CLI is installed. If not, follow https://docs.astral.sh/uv/ to install it.)
+uv sync --active
 
-The dataset path format is a HuggingFace id with the following name:
-```python
-"hirundo-io/bbq-{bias_type}-{either bias or unbias}-{multi-choice or free-text}"
+or 
+
+pip install -e .
```
[the remark this answers is not in the document]
Comment (tomerr-hirundo) [reply README.md:63]: sure, changed

File: examples/presets_customization.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        batch_size=32,
+        sample=False,
+        answer_tokens=128,
+        model_path_or_repo_id=model,
+        judge_batch_size=64,
+        judge_output_tokens=32,
+        judge_path_or_repo_id=judge_model,
+        use_4bit_judge=False,
+        results_dir=result_dir,
+    )
```
[the remark this answers is not in the document]
Comment (tomerr-hirundo) [reply examples/presets_customization.py:50]: I want to show people how to run the evaluation config as part of a full run

File: llm_behavior_eval/evaluation_utils/free_text_bias_evaluator.py [hunk 3]
```diff
@@ -212,14 +214,20 @@ def _match_llm_answers(
...
-        ):
+        if stereotyped_answers is None:
+            seq = zip(llm_answers, correct_answers, questions, strict=True)
+        else:
+            seq = zip(llm_answers, correct_answers, stereotyped_answers, questions, strict=True)
+        for items in seq:
+            if stereotyped_answers is None:
+                llm_ans, correct_ans, question = items  # type: ignore[misc]
+            else:
+                llm_ans, correct_ans, stereotyped_ans, question = items
```
[the remark this answers is not in the document]
Comment (tomerr-hirundo) [reply llm_behavior_eval/evaluation_utils/free_text_bias_evaluator.py:230]: ok can also do that, i tried to make it explicit but can change

[the remark this answers is not in the document]
Comment (tomerr-hirundo) [reply llm_behavior_eval/evaluation_utils/free_text_bias_evaluator.py:230]: note that if it is none i don't want to run the judge that checks for stereotypes, so the prompts still need to check it empty

File: llm_behavior_eval/evaluation_utils/free_text_hallu_evaluator.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+            input_ids = batch["test_input_ids"].to(self.model.device)
+            attention_mask = batch["test_attention_mask"].to(self.model.device)
+
+            input_texts = self.tokenizer.batch_decode(
+                input_ids, skip_special_tokens=True
+            )
+            gt_answers = self.tokenizer.batch_decode(
+                batch["gt_answers"], skip_special_tokens=True
+            )
+            out = cast("torch.nn.Module", self.model).generate(
```
[the remark this answers is not in the document]
Comment (tomerr-hirundo) [reply llm_behavior_eval/evaluation_utils/free_text_hallu_evaluator.py:86]: thanks

File: evaluate.py [hunk 5]
```diff
@@ -1,3 +1,4 @@
+import argparse
```
[the remark this answers is not in the document]
Comment (tomerr-hirundo) [reply evaluate.py]: ok

File: evaluate.py [hunk 6]
```diff
@@ -6,80 +7,114 @@
...
+def _behavior_presets(behavior: str) -> list[str]:
+    """
+    Map behavior presets to dataset identifiers (free‑text only).
+
+    New formats:
+    - BBQ: "bias:<bias_type>" or "unbias:<bias_type>"
+    - UNQOVER: "unqover:bias:<bias_type>" (UNQOVER does not support 'unbias')
+    - Hallucination retained for backward compatibility: "hallu" or "hallu-med"
+    """
+    behavior_parts = [p.strip().lower() for p in behavior.split(":")]
```
[the remark this answers is not in the document]
Comment (tomerr-hirundo) [reply evaluate.py]: ok

File: llm_behavior_eval/evaluation_utils/custom_dataset.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
+import logging
+from functools import partial
+from pathlib import Path
+from typing import Dict, List, cast
```
[the remark this answers is not in the document]
Comment (tomerr-hirundo) [reply llm_behavior_eval/evaluation_utils/custom_dataset.py:4]: thanks missed that

File: llm_behavior_eval/evaluation_utils/custom_dataset.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+) -> Dict[str, torch.Tensor]:
+    # 1) Column check
+    rows = [
+        dict(zip(examples_batch.keys(), vals))
+        for vals in zip(*examples_batch.values(), strict=True)
+    ]
+    # Validate minimally required fields only
+    for row in rows:
+        if not row.get("question") or not row.get("answer"):
+            raise ValueError("OpenText row must contain 'question' and 'answer'")
```
[the remark this answers is not in the document]
Comment (tomerr-hirundo) [reply llm_behavior_eval/evaluation_utils/custom_dataset.py:47]: thanks

File: llm_behavior_eval/evaluation_utils/custom_dataset.py [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+        answer_strings.append(answer_text)
+        if has_stereotype:
+            stereotyped_strings.append(stereotyped_text or "")
+
+    tokenize = partial(
+        tokenizer,
+        truncation=True,
+        padding="max_length",
+    )
+    enc_eval = tokenize(
```
[the remark this answers is not in the document]
Comment (tomerr-hirundo) [reply llm_behavior_eval/evaluation_utils/custom_dataset.py:68]: thanks

File: llm_behavior_eval/evaluation_utils/custom_dataset.py [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+
+        Args:
+            file_path: The local path or HuggingFace name of the dataset csv file.
+            dataset_type: The type of the dataset (e.g., BIAS or UNBIAS).
+        """
+        self.file_path = file_path
+        self.dataset_type = dataset_type
+        # Type "DatasetDict | Dataset | IterableDatasetDict | IterableDataset" is not assignable to declared type "DatasetDict"
+        # To ignore this error, since the loaded dataset is always a DatasetDict
+        # and we are using the "cast" function to ensure type safety.
```
[the remark this answers is not in the document]
Comment (tomerr-hirundo) [reply llm_behavior_eval/evaluation_utils/custom_dataset.py:115]: currently only accepting datasetdict so ill check for it. thanks

File: pyproject.toml [hunk 11]
```diff
@@ -27,15 +27,16 @@ keywords = [
...
+  "datasets==3.6.0",
+  "numpy==2.2.6",
+  "pandas==2.3.1",
+  "pydantic==2.11.7",
+  "pydantic_settings==2.10.1",
+  "torch==2.6.0",
+  "transformers==4.54.1",
+  "networkx==3.4.2",
+  "pytest==8.4.1",
+  "accelerate==1.10.0"
```
[the remark this answers is not in the document]
Comment (benglewis) [reply pyproject.toml]: I don't understand that at all.
1. If you are using `uv` and a `uv.lock` then the exact versions are locked anyway unless you actively choose to update them (https://docs.astral.sh/uv/concepts/projects/sync/#syncing-development-dependencies)
2. If you are not using `uv`, I still would much prefer that we find some way to actually set the requirements correctly, e.g. if certain `triton` / `transformers` versions don't play nicely together, it would be nice if we can make this `pyproject.toml` state that rather than specific versions of everything.
3. Having fixed versions here will just make it much more fragile and harder to upgrade versions later when we have to for vulnerability reasons.

File: examples/presets_customization.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tomerr-hirundo) [reply examples/presets_customization.py:50]: Yeah if they want to test a model not supported by the list of models

[the remark this answers is not in the document]
Comment (tomerr-hirundo) [reply examples/presets_customization.py:50]: I mainly support llama and gemma models currently

[the remark this answers is not in the document]
Comment (benglewis) [reply examples/presets_customization.py:50]: Got it

File: pyproject.toml (hunk 11 above)
[the remark this answers is not in the document]
Comment (benglewis) [reply pyproject.toml]: 1. Honestly, I am not sure if we have to support `pip`, but, if you want to, we can use [`pylock.toml`](https://packaging.python.org/en/latest/specifications/pylock-toml/), the Python standard for _lockfiles_ which `uv` can now already [export to](https://github.com/astral-sh/uv/issues/12584). Unfortunately, pip cannot install from that yet. But you can also export to `requirements.txt` and so we can keep a copy of that for `pip`. Hopefully the TOML version will be supported in the future.
2. What is the exact connection between them?
3. Users don't install it via our repo anyway... they use the Python package. I agree that we need it to work. How does this approach look to you? https://chatgpt.com/share/e/68ab05c5-db30-8011-9d72-1748a333641a

File: llm_behavior_eval/evaluate.py [hunk 12]
```diff
@@ -66,28 +71,21 @@ def _behavior_presets(behavior: str) -> list[str]:
...
-        "--model",
-        required=True,
-        help="Model repo id or path, e.g. meta-llama/Llama-3.1-8B-Instruct",
-    )
-    parser.add_argument(
-        "--behavior",
-        required=True,
-        help=(
-            "Behavior preset. BBQ: 'bias:<type>' or 'unbias:<type>'; "
-            "UNQOVER: 'unqover:bias:<type>'; Hallucination: 'hallu' | 'hallu-med'"
```
Comment (benglewis) [inline llm_behavior_eval/evaluate.py]: Any particular reason that you remove this from the help string?

File: llm_behavior_eval/evaluation_utils/custom_dataset.py [hunk 13]
```diff
@@ -44,11 +43,11 @@ def open_text_preprocess_function(
     # Validate minimally required fields only
     for row in rows:
         if not row.get("question") or not row.get("answer"):
-            raise ValueError("OpenText row must contain 'question' and 'answer'")
+            raise ValueError("FreeText row must contain 'question' and 'answer'")
```
Comment (benglewis) [inline llm_behavior_eval/evaluation_utils/custom_dataset.py:46]:
```suggestion
            raise ValueError("Free text row must contain 'question' and 'answer'")
```

File: pyproject.toml [hunk 14]
```diff
@@ -36,6 +39,7 @@ dependencies = [
   "transformers==4.54.1",
   "networkx==3.4.2",
   "pytest==8.4.1",
+  "typer==0.16.1",
```
Comment (benglewis) [inline pyproject.toml:42]: See previous discussion. I really really prefer not to lock requirement versions where possible

File: llm_behavior_eval/evaluate.py (hunk 12 above)
Comment (tomerr-hirundo) [reply llm_behavior_eval/evaluate.py]: I updated the Annotated comment with it

File: pyproject.toml (hunk 14 above)
Comment (tomerr-hirundo) [reply pyproject.toml:42]: I really don't
I don't know how how to solve this comment

Comment (tomerr-hirundo) [reply pyproject.toml:42]: @mishana?

File: pyproject.toml [hunk 15]
```diff
@@ -2,6 +2,9 @@
 requires = ["setuptools>=64", "wheel"]
 build-backend = "setuptools.build_meta"
 
+[project.scripts]
+llm-behavior-eval = "llm_behavior_eval.evaluate:app"
```
[the remark this answers is not in the document]
Comment (tomerr-hirundo) [reply pyproject.toml:6]: very nice! it works

File: llm_behavior_eval/evaluation_utils/free_text_hallu_evaluator.py [hunk 16]
```diff
@@ -133,7 +135,32 @@ def _grade_batch(
...
+            serializable = [
+                {
+                    "input_texts": g.input_texts,
+                    "gt_answers": g.gt_answers,
+                    "answers": g.answers,
+                }
+                for g in generations
+            ]
+            with open(generations_path, "w") as f:
+                json.dump(serializable, f, indent=2)
```
[the remark this answers is not in the document]
Comment (benglewis) [reply llm_behavior_eval/evaluation_utils/free_text_hallu_evaluator.py:163]: Never mind, I notice that you did mostly the same as this in LLM-21. Closing

File: README.md [hunk 17]
```diff
@@ -2,67 +2,135 @@
...
-Make sure you have Python 3.10+ installed, then install dependencies:
+Make sure you have Python 3.10+ installed, then set up a virtual environment and install dependencies with `uv`:
 
 ```bash
-git clone https://github.com/your-org/bias-evaluation.git
-cd bias-evaluation
-pip install -e .
+# 1) Clone and enter the repository
+git clone https://github.com/hirundo-io/llm-behavior-eval.git
+cd llm-behavior-eval
```
[the remark this answers is not in the document]
Comment (tomerr-hirundo) [reply README.md:57]: Thanks, i deleted this part

File: llm_behavior_eval/evaluate.py [hunk 18]
```diff
@@ -41,15 +41,21 @@ def _behavior_presets(behavior: str) -> list[str]:
...
+    #   - bias_type can be a concrete type or 'all'
     if len(behavior_parts) == 2:
         kind, bias_type = behavior_parts
         if kind not in BIAS_KINDS:
             raise ValueError("For BBQ use 'bias:<bias_type>' or 'unbias:<bias_type>'")
         from llm_behavior_eval.evaluation_utils.enums import BBQ_BIAS_TYPES
 
+        if bias_type == "all":
+            return [
+                f"hirundo-io/bbq-{bt}-{kind}-free-text" for bt in sorted(BBQ_BIAS_TYPES)
```
[the remark this answers is not in the document]
Comment (tomerr-hirundo) [reply llm_behavior_eval/evaluate.py:55]: thanks

File: llm_behavior_eval/evaluation_utils/base_evaluator.py [hunk 19]
```diff
@@ -118,10 +138,9 @@ def save_results(
 
         output_responses = output_dir / "responses.json"
         output_metrics = output_dir / "metrics.csv"
-        error = 1 - accuracy
         results = pd.DataFrame(
             {
-                "Error": [error],
+                "Metric Value": [metric_value],
```
[the remark this answers is not in the document]
Comment (tomerr-hirundo) [reply llm_behavior_eval/evaluation_utils/base_evaluator.py:143]: Agreed

File: requirements.txt [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+exceptiongroup==1.3.0
+filelock==3.18.0
+frozenlist==1.6.0
+fsspec==2025.3.0
+hf-xet==1.1.7
+huggingface-hub==0.34.4
+idna==3.10
+iniconfig==2.1.0
+jinja2==3.1.6
+llm-behavior-eval @ file:///home/ubuntu/llm-behavior-eval
```
[the remark this answers is not in the document]
Comment (benglewis) [reply requirements.txt:23]: Nope. It is intentional

File: llm_behavior_eval/evaluation_utils/base_evaluator.py [hunk 21]
```diff
@@ -220,9 +218,7 @@ def infer_bias_label_from_slug(slug: str) -> str:
             return slug
 
         bias_label = infer_bias_label_from_slug(dataset_slug)
-        brief_df = pd.DataFrame(
-            {"Bias Type": [bias_label], "Metric Value": [metric_value]}
-        )
+        brief_df = pd.DataFrame({"Dataset": [bias_label], "Metric value (accuracy for unbias & hallucinations / error otherwise)": [to_report_score]})
```
Comment (benglewis) [inline llm_behavior_eval/evaluation_utils/base_evaluator.py:221]: Why can't you also change this to use the correct name for the column here? It just seems really strange the whole "Metric value (...)" bit

**Review by benglewis — APPROVED** [review]
One question, but otherwise LGTM

File: llm_behavior_eval/evaluation_utils/base_evaluator.py (hunk 21 above)
Comment (tomerr-hirundo) [reply llm_behavior_eval/evaluation_utils/base_evaluator.py:221]: Changed to the column name, so we could have accuracy and error rate columns but i guess that's only two columns so that's fine

**Review by benglewis — APPROVED** [review]
LGTM
